k-distribution/k-tutorial/1_basic/22_proofs/README.md

```diff
@@ -1 +1,444 @@
+statement, which has a symbolic condition (`A < B`), updating the value of the
+program variable `$c` in both branches. The specification we give states that
+the `if` statement terminates, with `$a` and `$b` updated, respectively, to `A`
+and `B`, and `$c` updated to *some* symbolic integer value `?C`. Via the
+`ensures` clause, which is used to specify additional constraints that hold
+after execution, we also state that this existentially quantified `?C` equals
+either `A` or `B`.
+
+Add the productions declaring `$b` and `$c` as tokens to the
+`LESSON-22-SPEC-SYNTAX` module, the claim to the `LESSON-22-SPEC` module, run
+the K prover again, and observe the output, which should not be `#Top` this
+time. This means that K was not able to prove the claim, and we now need to
+understand why. We do so by examining the output, which should look as follows:
+
+```
+    (InfoReachability) while checking the implication:
+    The configuration's term unifies with the destination's term,
+    but the implication check between the conditions has failed.
+
+  #Not (
+    #Exists ?C . {
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- ?C:Int ]
+      #Equals
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    }
+  #And
+    {
+      true
+    #Equals
+      ?C ==Int A orBool ?C ==Int B
+    }
+  )
+#And
+  <generatedTop>
+    <k>
+      _DotVar1
+    </k>
+    <store>
+      STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    </store>
+    <funcs>
+      _Gen3
+    </funcs>
+    <stack>
+      _Gen5
+    </stack>
+  </generatedTop>
+#And
+  {
+    true
+  #Equals
+    A <Int B
+  }
+```
+
+This output starts with a message telling us at which point the proof failed,
+followed by the final state, which consists of three parts: some negative
+Matching Logic (ML) constraints, the final configuration (`<generatedTop> ...
+</generatedTop>`), and some positive ML constraints.
+
+First, we examine the message:
+
+```
+(InfoReachability) while checking the implication:
+The configuration's term unifies with the destination's term,
+but the implication check between the conditions has failed.
+```
+
+which tells us that the *structure* of the final configuration is as expected,
+but that some of the associated constraints cannot be proven. We next look at
+the final configuration, in which the relevant item is the `<store> ...
+</store>` cell, because it is the only one that we are reasoning about. By
+inspecting its contents:
+
+```
+STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+```
+
+we see that we should be within the constraints of the `ensures`, since the
+value of `$c` in the store equals `B` in this branch. We next examine the
+negative and positive constraints of the output and, more often than not, the
+goal is to instruct K how to use the information from the final configuration
+and the positive constraints to falsify one of the negative constraints. This
+is done through *simplifications*.
```

Comment:
When I went through this I thought the parts about negative constraints could use more explanation, like how the negative constraints are generated or something, or more on their meaning. I just found it confusing that at least in this example the negative constraint is just the negation of something that is actually true but that the prover wasn't able to prove, whereas other constraints are generated by taking branches where it is assumed that constraint holds.